### PR TITLE
Modify AD graph commands to sync credential management in app/sp and other fixes

### PR DESCRIPTION
**Pull request by shuagarw**
The content to be added to Changelog is as follows:
* Simplify ServicePrincipal creation when there is no corresponding existing application.
* Simplify credential creation for new/update app and servicePrincipal

### CODE REVIEW

File: lib/commands/arm/ad/ad.application._js [hunk 1]
```diff
@@ -36,13 +36,11 @@ exports.init = function (cli) {
     .usage('[options] <name> <home-page> <identifier-uris>')
     .option('-n --name <name>', $('the display name for the application'))
     .option('-m --home-page <home-page>', $('the URL to the application homepage'))
-    .option('-a --available', $('indicates if the application will be available to other tenants'))
-    .option('-p --password <password>', $('the value for the password credential associated with the application that will be valid for one year by default'))
     .option('-i --identifier-uris <identifier-uris>', $('the comma-delimitied URIs that identify the application'))
+    .option('-l --available <available>', $('the value specifying whether the application is a single tenant or a multi-tenant. "true" if the application will be available to other tenants, "false" if its a single tenant application. Default is false.'))
```
Comment (amarzavery) [inline lib/commands/arm/ad/ad.application._js]: why is this changed to -l ?

Comment (amarzavery) [inline lib/commands/arm/ad/ad.application._js]: This can be made much simpler. Dont accept the value. Then it will be a boolean flag. Then you dont need to verify that it is string true or string false. If the user does not specify the switch then it is false else it is true.

**.option('-l --available', . . . );**

File: lib/commands/arm/ad/ad.application._js [hunk 2]
```diff
@@ -58,14 +56,13 @@ exports.init = function (cli) {
     .description($('Updates the properties of the created Active Directory application'))
     .option('-o --objectId <objectId>', $('the object id of the application to update. Either provide objectId or applicationId. objectId is more preferable.'))
     .option('-a --applicationId <applicationId>', $('the applicationId of the application to update. If you have created a service principal using this application then this will be the spn of the service principal.'))
+    .option('-l --available <available>', $('the new value specifying whether the application is a single tenant or a multi-tenant. "true" if the application will be available to other tenants, "false" if its a single tenant application.'))
```
Comment (amarzavery) [inline lib/commands/arm/ad/ad.application._js]: same as above

File: lib/commands/arm/ad/ad.application._js [hunk 3]
```diff
@@ -85,13 +82,24 @@ exports.init = function (cli) {
...
+      if (options.available) {
+        if (options.available.toLowerCase() === 'true') {
+          options.available = true;
+        }
+        else if (options.available.toLowerCase() === 'false') {
+          options.available = false;
+        } else {
+          throw new Error($('Invalid value for --available parameter. Please specify "true" for a multi-tenant application or "false" for a single-tenant application.'));
+        }
+      }
```
Comment (amarzavery) [inline lib/commands/arm/ad/ad.application._js]: The entire above logic goes away. Simply pass options.available to the sdk

File: lib/commands/arm/ad/ad.servicePrincipals._js [hunk 4]
```diff
@@ -93,35 +93,78 @@ exports.init = function (cli) {
...
-    .option('-m --home-page <home-page>', $('the URL to the application homepage'))
-    .option('-b --available', $('indicates if the application will be available to other tenants'))
     .option('-p --password <password>', $('the value for the password credential associated with the application that will be valid for one year by default'))
-    .option('-i --identifier-uris <identifier-uris>', $('the comma-delimitied URIs that identify the application'))
-    .option('-r --reply-urls <reply-urls>', $('the comma-delimitied application reply urls'))
-    .option('--key-value <key-value>', $('the value for the key credentials associated with the application that will be valid for one year by default'))
-    .option('--key-type <key-type>', $('the type of the key credentials associated with the application. Acceptable values are AsymmetricX509Cert, Password and Symmetric'))
-    .option('--key-usage <key-usage>', $('the usage of the key credentials associated with the application. Acceptable values are Sign and Verify'))
+    .option('--cert-value <cert-value>', $('the value of the "asymmetric" credential type. It represents the base 64 encoded certificate'))
```
Comment (amarzavery) [inline lib/commands/arm/ad/ad.servicePrincipals._js]: What if the user provides password and cert value? Are they mutually exclusive? If yes then we should validate that and throw the right error

File: lib/commands/arm/ad/ad.servicePrincipals._js [hunk 5]
```diff
@@ -139,13 +182,8 @@ exports.init = function (cli) {
...
-    .option('-m --home-page <home-page>', $('the new URL to the application homepage'))
     .option('-p --password <password>', $('new value for the password credential associated with the application that will be valid for one year by default'))
-    .option('-i --identifier-uris <identifier-uris>', $('new comma-delimitied URI values that identify the application'))
-    .option('-r --reply-urls <reply-urls>', $('new comma-delimitied application reply urls'))
-    .option('--key-value <key-value>', $('new value for the key credentials associated with the application that will be valid for one year by default'))
-    .option('--key-type <key-type>', $('new type of the key credentials associated with the application. Acceptable values are AsymmetricX509Cert, Password and Symmetric. Default value: AsymmetricX509Cert'))
-    .option('--key-usage <key-usage>', $('new usage of the key credentials associated with the application. Acceptable values are Sign and Verify. Default value: Verify'))
+    .option('--cert-value <cert-value>', $('new value of the "asymmetric" credential type. It represents the base 64 encoded certificate'))
```
Comment (amarzavery) [inline lib/commands/arm/ad/ad.servicePrincipals._js]: same as above. password and cert-value mutually exclusive ?

File: lib/commands/arm/ad/ad.application._js [hunk 6]
```diff
@@ -58,14 +56,13 @@ exports.init = function (cli) {
...
     .option('-i --identifier-uris <identifier-uris>', $('new comma-delimitied URI values that identify the application'))
     .option('-r --reply-urls <reply-urls>', $('new comma-delimitied application reply urls'))
-    .option('--key-value <key-value>', $('new value for the key credentials associated with the application that will be valid for one year by default'))
-    .option('--key-type <key-type>', $('new type of the key credentials associated with the application. Acceptable values are AsymmetricX509Cert, Password and Symmetric. Default value: AsymmetricX509Cert'))
-    .option('--key-usage <key-usage>', $('new usage of the key credentials associated with the application. Acceptable values are Sign and Verify. Default value: Verify'))
+    .option('-p --password <password>', $('new value for the password credential associated with the application that will be valid for one year by default'))
+    .option('--cert-value <cert-value>', $('new value of the "asymmetric" credential type. It represents the base 64 encoded certificate.'))
```
Comment (amarzavery) [inline lib/commands/arm/ad/ad.application._js]: same as above. password and cert-value mutually exclusive ?

File: lib/commands/arm/ad/ad.application._js [hunk 7]
```diff
@@ -36,13 +36,11 @@ exports.init = function (cli) {
...
+    .option('-l --available <available>', $('the value specifying whether the application is a single tenant or a multi-tenant. "true" if the application will be available to other tenants, "false" if its a single tenant application. Default is false.'))
     .option('-r --reply-urls <reply-urls>', $('the comma-delimitied application reply urls'))
-    .option('--key-value <key-value>', $('the value for the key credentials associated with the application that will be valid for one year by default'))
-    .option('--key-type <key-type>', $('the type of the key credentials associated with the application. Acceptable values are AsymmetricX509Cert, Password and Symmetric'))
-    .option('--key-usage <key-usage>', $('the usage of the key credentials associated with the application. Acceptable values are Sign and Verify'))
+    .option('-p --password <password>', $('the value for the password credential associated with the application that will be valid for one year by default'))
+    .option('--cert-value <cert-value>', $('the value of the "asymmetric" credential type. It represents the base 64 encoded certificate.'))
```
Comment (amarzavery) [inline lib/commands/arm/ad/ad.application._js]: password and cert-value mutually exclusive If yes then throw an error. Mention this in the help as well?

File: lib/commands/arm/ad/ad.servicePrincipals._js [hunk 8]
```diff
@@ -259,7 +305,7 @@ exports.init = function (cli) {
     .description($('Deletes Active Directory service principal.'))
     .usage('[options] <objectId>')
     .option('-o --objectId <objectId>', $('the object id of the service principal to delete'))
-    .option('-p --preserve-application', $('Default value: false. If you do not want to delete the underlying application then set this flag.'))
+    .option('-d --delete-application', $('Default value: false. If you want to delete the underlying application then set this flag.'))
```
Comment (amarzavery) [inline lib/commands/arm/ad/ad.servicePrincipals._js]: see, you correctly used the boolean flag over here

File: lib/commands/arm/ad/adUtils._js [hunk 9]
```diff
@@ -38,8 +38,20 @@ exports.createApplication = function (cli, name, homePage, identifierUris, optio
     return cli.missingArgument('identifier-uris');
   }
 
-  if (options.password && options.keyValue) {
-    throw new Error($('specify either --password or --key-value, but not both.'));
+  if (options.password && options.certValue) {
+    throw new Error($('specify either --password or --cert-value, but not both.'));
+  }
+
+  var available = false;
```
Comment (amarzavery) [inline lib/commands/arm/ad/adUtils._js]: All of this will go away

File: lib/commands/arm/ad/adUtils._js [hunk 10]
```diff
@@ -38,8 +38,20 @@ exports.createApplication = function (cli, name, homePage, identifierUris, optio
     return cli.missingArgument('identifier-uris');
   }
 
-  if (options.password && options.keyValue) {
-    throw new Error($('specify either --password or --key-value, but not both.'));
+  if (options.password && options.certValue) {
```
Comment (amarzavery) [inline lib/commands/arm/ad/adUtils._js]: Is this happening in all the methods?

File: lib/commands/arm/ad/ad.application._js (hunk 1 above)
Comment (shuagarw) [inline lib/commands/arm/ad/ad.application._js]: Yes, I intentionally did it this way, because this parameter is also specified in set command. Where it wont be possible to distinguish that you didn't want to update this param or you want to update with false value. And for symmetry changed here as well. Let me know your thoughts,

File: lib/commands/arm/ad/ad.application._js (hunk 2 above)
Comment (shuagarw) [inline lib/commands/arm/ad/ad.application._js]: As above.

File: lib/commands/arm/ad/ad.application._js (hunk 3 above)
Comment (shuagarw) [inline lib/commands/arm/ad/ad.application._js]: As above.

File: lib/commands/arm/ad/ad.servicePrincipals._js (hunk 4 above)
Comment (shuagarw) [inline lib/commands/arm/ad/ad.servicePrincipals._js]: They should not be mutually exclusive. It was like that, so I didnt change it. But its better we change it now rather keeping it the same way.

File: lib/commands/arm/ad/ad.servicePrincipals._js (hunk 5 above)
Comment (shuagarw) [inline lib/commands/arm/ad/ad.servicePrincipals._js]: Will fix it. They are not mutually exclusive.

File: lib/commands/arm/ad/ad.servicePrincipals._js (hunk 4 above)
Comment (shuagarw) [inline lib/commands/arm/ad/ad.servicePrincipals._js]: As discussed, made all the command cert and password creds mutually exclusive.

File: lib/commands/arm/ad/ad.servicePrincipals._js (hunk 5 above)
Comment (shuagarw) [inline lib/commands/arm/ad/ad.servicePrincipals._js]: Same as above.

File: lib/commands/arm/ad/ad.application._js (hunk 6 above)
Comment (shuagarw) [inline lib/commands/arm/ad/ad.application._js]: same as above.

File: lib/commands/arm/ad/ad.application._js (hunk 7 above)
Comment (shuagarw) [inline lib/commands/arm/ad/ad.application._js]: same as above.

File: lib/commands/arm/ad/adUtils._js (hunk 9 above)
Comment (shuagarw) [inline lib/commands/arm/ad/adUtils._js]: As discussed this is required.

File: lib/commands/arm/ad/adUtils._js (hunk 10 above)
Comment (shuagarw) [inline lib/commands/arm/ad/adUtils._js]: Yes, fixed to enforce check in all methods.

File: lib/commands/arm/ad/ad.application._js (hunk 3 above)
Comment (stefmilmsft) [inline lib/commands/arm/ad/ad.application._js]: IIRC the directory allows you to mix & match passwords & certificates. Are we consciously limiting this in the client bits?
